### PR TITLE
Preliminaries for scroll anchoring

### PR DESCRIPTION
**Pull request by abarth**
These patches are warmups for adding support for scroll anchors.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @HansMuller

File: packages/flutter/lib/src/material/tabs.dart [hunk 1]
```diff
@@ -922,12 +922,12 @@ class _TabBarViewState extends PageableListState<TabBarView> implements TabBarSe
...
 
-  Future fling(Offset scrollVelocity) {
+  Future fling(double scrollVelocity) {
     if (_selection == null || _selection.valueIsChanging)
       return new Future.value();
 
-    if (scrollVelocity.dx.abs() > _kMinFlingVelocity) {
-      final int selectionDelta = scrollVelocity.dx > 0 ? -1 : 1;
+    if (scrollVelocity.abs() > _kMinFlingVelocity) {
+      final int selectionDelta = scrollVelocity.sign.round();
```
Comment (Hixie) [inline packages/flutter/lib/src/material/tabs.dart]: truncate() is probably more precisely the semantic we're going for here than round()

File: packages/flutter/lib/src/rendering/grid.dart [hunk 2]
```diff
@@ -338,6 +340,11 @@ class RenderGrid extends RenderVirtualViewport<GridParentData> {
     _delegate = newDelegate;
   }
 
+  void set scrollDirection(Axis value) {
+    assert(value == Axis.vertical && "RenderGrid doesn't yet support horizontal scrolling." is String);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/grid.dart]:
```dart
assert(() {
  if (value != Axis.vertical)
    throw WidgetError('RenderGrid doesn\'t yet support horizontal scrolling.');
  return true;
});
```

**Comment by Hixie** [comment]
lgtm

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 3]
```diff
@@ -66,13 +68,72 @@ class RenderViewport extends RenderBox with RenderObjectWithChildMixin<RenderBox
...
+  }
+
+}
+
+/// A render object that's bigger on the inside.
+///
+/// The child of a viewport can layout to a larger size than the viewport
+/// itself. If that happens, only a portion of the child will be visible through
+/// the viewport. The portion of the child that is visible is controlled by the
+/// scroll offset.
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/viewport.dart]: paintOffset

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 4]
```diff
@@ -66,13 +68,72 @@ class RenderViewport extends RenderBox with RenderObjectWithChildMixin<RenderBox
...
+///
+/// Viewport is the core scrolling primitive in the system, but it can be used
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/viewport.dart]: Isn't Scrollable the core scrolling primitive? I guess Viewport is just a clipping building block for Scrollables and others.

Comment (abarth) [inline packages/flutter/lib/src/rendering/viewport.dart]: Yeah, I'll remove this comment.  It's out-of-date.

**Pull request merged**